File: C/0012.c
#include <stdio.h>
#include <math.h>

// Function to calculate the number of divisors of a given number
int countDivisors(int num) {
    int count = 0;
    int sqrtNum = (int) sqrt(num);
    for (int i = 1; i <= sqrtNum; i++) {
        if (num % i == 0) {
            if (i * i == num) {
                count++; // i and num/i are the same
            } else {
                count += 2; // i and num/i are different
            }
        }
    }
    return count;
}

// Function to find the first triangle number with more than 'divisorLimit' divisors
int findTriangleNumberWithDivisors(int divisorLimit) {
    int n = 1;
    int triangleNumber = 0;

    while (1) {
        triangleNumber += n;
        int numDivisors = countDivisors(triangleNumber);

        if (numDivisors > divisorLimit) {
            return triangleNumber;
        }

        n++;
    }
}

int main() {
    int divisorLimit = 500;
    int result = findTriangleNumberWithDivisors(divisorLimit);
    printf("The first triangle number with more than %d divisors is %d\n", divisorLimit, result);
    return 0;
}

File: TypeScript/0012.ts
function firstTriangleNumberWithDivisors(minDivisors: number): number {
    let n = 1;
    let triangleNumber = 1;

    while (true) {
        // Get the number of divisors for the current triangle number
        const divisors = countDivisors(triangleNumber);

        if (divisors > minDivisors) {
            return triangleNumber;
        }

        // Generate next triangle number
        n++;
        triangleNumber += n;
    }
}

function countDivisors(num: number): number {
    if (num === 1) return 1;

    let count = 1;
    let remaining = num;

    // Handle 2 separately
    if (remaining % 2 === 0) {
        let exponent = 0;
        while (remaining % 2 === 0) {
            remaining /= 2;
            exponent++;
        }
        count *= (exponent + 1);
    }

    // Check odd divisors up to sqrt(remaining)
    for (let i = 3; i <= Math.sqrt(remaining); i += 2) {
        if (remaining % i === 0) {
            let exponent = 0;
            while (remaining % i === 0) {
                remaining /= i;
                exponent++;
            }
            count *= (exponent + 1);
        }
    }

    // If remaining is a prime number greater than 2
    if (remaining > 2) {
        count *= 2;
    }

    return count;
}

const minDivisors = 500;
const result = firstTriangleNumberWithDivisors(minDivisors);
console.log(`The first triangle number with over ${minDivisors} divisors is: ${result}`);